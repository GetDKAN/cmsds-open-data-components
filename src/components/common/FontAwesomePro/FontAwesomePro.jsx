import React, { useRef, useEffect } from 'react'
import PropTypes from 'prop-types'
import svgIndex from '../../../assets/icons'

const FontAwesomePro = (props) => {
  const {
    'aria-hidden': ariaHidden,
    fill = '#323A45',
    height = 16,
    icon = 'long-arrow-right',
    size,
    width = 16
  } = props

  // Set fill on path
  const ref = useRef()
  useEffect(() => {
    if (fill && ref.current && ref.current.firstElementChild) {
      ref.current.firstElementChild.setAttribute('fill', fill)
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [icon])

  // Set default svg attribtues
  // const iconAttr = { /////////////////
  //   'aria-hidden': ariaHidden,
  //   className: `fa fa-${icon}`,
  //   focusable: 'false',
  //   height: size || height,
  //   role: 'img',
  //   width: size || width
  // }

  // Get svgIndex export array values
  const svgList = svgIndex.values() ////////

  // Loop through svgIndex export array
  // const iconMap = (ico) => { //////////////
  //   for (const svg of svgList) {
  //     const key = Object.keys(svg)[0]
  //     if (key === ico) {
  //       // Get value from svgIndex, which is a ReactComponent
  //       const Value = Object.values(svg)[0]
  //       return <Value {...iconAttr} ref={ref} aria-label={icon.split('-').join(' ')} />
  //     }
  //   }
  // }

  // return ( ///////
  //   iconMap(icon)
  // )
}

FontAwesomePro.propTypes = {
  /**
   * Hide icon from screen reader
   */
  'aria-hidden': PropTypes.bool,
  /**
   * Icon color code string (hex, rgb, string, etc)
   */
  fill: PropTypes.string,
  /**
   * SVG outer pixel height - Overwritten by 'size' prop if provided
   */
  height: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /**
   * FontAwesome icon name
   */
  icon: PropTypes.string,
  /**
   * SVG pixel size - Applies to height and width equally
   */
  size: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /**
   * SVG outer pixel width - Overwritten by 'size' prop if provided
   */
  width: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
}
FontAwesomePro.displayName = 'FontAwesomePro'
export default FontAwesomePro
