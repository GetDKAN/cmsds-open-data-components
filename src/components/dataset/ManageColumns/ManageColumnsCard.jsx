import React from 'react'
import PropTypes from 'prop-types'
import { useDrag, useDrop } from 'react-dnd'

const ItemTypes = {
  CARD: 'card'
}

const ManageColumnsCard = ({ id, index, moveCard, children }) => {
  const ref = React.useRef(null)
  const [, drop] = useDrop({
    accept: ItemTypes.CARD,
    hover (item, monitor) {
      if (!ref.current) {
        return
      }

      const dragIndex = item.index
      const hoverIndex = index

      // Don't replace items with themselves
      if (dragIndex === hoverIndex) {
        return
      }

      // Determine rectangle on screen
      const hoverBoundingRect = ref.current.getBoundingClientRect()

      // Get vertical middle
      const hoverMiddleY =
        (hoverBoundingRect.bottom - hoverBoundingRect.top) / 2

      // Determine mouse position
      const clientOffset = monitor.getClientOffset()

      // Get pixels to the top
      const hoverClientY = clientOffset.y - hoverBoundingRect.top

      // Only perform the move when the mouse has crossed half of the items height
      // When dragging downwards, only move when the cursor is below 50%
      // When dragging upwards, only move when the cursor is above 50%
      // Dragging downwards
      if (dragIndex < hoverIndex && hoverClientY < hoverMiddleY) {
        return
      }

      // Dragging upwards
      if (dragIndex > hoverIndex && hoverClientY > hoverMiddleY) {
        return
      }

      // Time to actually perform the action
      moveCard(dragIndex, hoverIndex)

      // Note: we're mutating the monitor item here!
      // Generally it's better to avoid mutations,
      // but it's good here for the sake of performance
      // to avoid expensive index searches.
      item.index = hoverIndex
    }
  })

  const [{ isDragging }, drag] = useDrag({
    item: { id, index },
    type: ItemTypes.CARD,
    collect: (monitor) => ({
      isDragging: monitor.isDragging()
    })
  })

  const opacity = isDragging ? 0 : 1

  drag(drop(ref))

  return (
    <li key={id} ref={ref} style={{ opacity }}>
      {children}
    </li>
  )
}

ManageColumnsCard.propTypes = {
  /**
   * Used for `key` attribute on wrapping element
   */
  id: PropTypes.string.isRequired,
  /**
   * Index ID used for drag/drop logic
   */
  index: PropTypes.number.isRequired,
  /**
   * Child elements inside drag/drop element
   */
  children: PropTypes.node.isRequired,
  /**
   * Callback function fired when item is moved.
   * `dragIndex` and `hoverIndex` are passed as arguments
   */
  moveCard: PropTypes.func.isRequired
}

ManageColumnsCard.displayName = 'ManageColumnsCard'
export default ManageColumnsCard
