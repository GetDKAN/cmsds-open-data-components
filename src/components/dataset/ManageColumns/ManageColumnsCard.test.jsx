import { render } from '@testing-library/react'
import ManageColumnsCard from './ManageColumnsCard'
import FontAwesomePro from '../../common/FontAwesomePro/FontAwesomePro'
import { DndProvider } from 'react-dnd'
import { HTML5Backend } from 'react-dnd-html5-backend'
import { Choice } from '@cmsgov/design-system'

/*
  We can't fully test this components included logic because DndProvider actually lives in ManageColumns.js 
  and this component only yields a single card and we'd need the full DnD list in order to test all functionality.
  We have unit tests for ManageColumns.js that test reordering the DnD list.
*/

const mockMoveCard = vi.fn()
const componentArgs = {
  id: 'col1',
  index: 0,
  children: (
    <>
      <Choice
        id='col1'
        type='checkbox'
        checked
        name='col1'
        value=''
        label='col1'
      />
      <div className='manage-columns-reorder'>
        <FontAwesomePro icon='sort' customClass='manage-columns-reorder-icon' />
      </div>
    </>
  ),
  moveCard: mockMoveCard
}

const renderComponent = (args = componentArgs) => render (
  <DndProvider backend={HTML5Backend}>
    <ManageColumnsCard {...args} />
  </DndProvider>
)

describe('ManageColumnsCard component.', () => {
  it('Matches snapshot.', () => {
    const renderedManageColumnsCard = renderComponent()

    expect(renderedManageColumnsCard.asFragment()).toMatchSnapshot()
  })
})
