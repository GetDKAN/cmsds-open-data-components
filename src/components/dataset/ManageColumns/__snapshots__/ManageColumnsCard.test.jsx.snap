// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ManageColumnsCard component. > Matches snapshot. 1`] = `
<DocumentFragment>
  <li
    draggable="true"
    style="opacity: 1;"
  >
    <div>
      <div
        class="ds-c-choice-wrapper"
      >
        <input
          aria-describedby="col1__error"
          checked=""
          class="ds-c-choice"
          id="col1"
          name="col1"
          type="checkbox"
          value=""
        />
        <label
          class="ds-c-label"
          for="col1"
          id="col1__label"
        >
          col1
        </label>
        <p
          aria-atomic="true"
          aria-live="assertive"
          class="ds-c-inline-error"
          id="col1__error"
        />
      </div>
    </div>
    <div
      class="manage-columns-reorder"
    >
      <svg
        aria-label="sort"
        class="fa fa-sort"
        focusable="false"
        height="16"
        role="img"
        viewBox="0 0 320 512"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41zm255-105L177 64c-9.4-9.4-24.6-9.4-33.9 0L24 183c-15.1 15.1-4.4 41 17 41h238c21.4 0 32.1-25.9 17-41z"
          fill="#323A45"
        />
      </svg>
    </div>
  </li>
</DocumentFragment>
`;
